### PR TITLE
[metal] Add specialization constants

### PR DESCRIPTION
**Pull request by grovesNL**
- Removed push constants from quad example so Metal can run again
- Increased minimum Metal version to 1.2 (necessary for function constants), at least for now
- Added support for specialization constants

WIP because it requires https://github.com/gfx-rs/metal-rs/pull/21 to be published first.

### CODE REVIEW

File: src/backend/metal/src/device.rs [hunk 1]
```diff
@@ -45,6 +42,24 @@ fn gen_parse_error(err: SpirvErrorCode) -> ShaderError {
     ShaderError::CompilationFailed(msg)
 }
 
+fn create_function_constants(specialization: &[pso::Specialization]) -> metal::FunctionConstantValues {
+    let constants_raw = metal::FunctionConstantValues::new();
+    use std::os::raw::c_void;
+    for constant in specialization {
+        unsafe {
+            let tv = match constant.value {
+                pso::Constant::Bool(ref v) => (MTLDataType::Bool, v as *const _ as *const c_void),
```
Comment (msiglreith) [inline src/backend/metal/src/device.rs]: nit: `as *const _ as *const _` should also be enough

File: src/backend/metal/src/device.rs [hunk 2]
```diff
@@ -45,6 +42,24 @@ fn gen_parse_error(err: SpirvErrorCode) -> ShaderError {
     ShaderError::CompilationFailed(msg)
 }
 
+fn create_function_constants(specialization: &[pso::Specialization]) -> metal::FunctionConstantValues {
+    let constants_raw = metal::FunctionConstantValues::new();
+    use std::os::raw::c_void;
+    for constant in specialization {
+        unsafe {
+            let tv = match constant.value {
```
Comment (msiglreith) [inline src/backend/metal/src/device.rs]: Could unpack the tuple already here.

File: src/backend/metal/src/device.rs [hunk 3]
```diff
@@ -254,7 +267,7 @@ impl Device {
         debug!("SPIRV-Cross generated shader:\n{}", shader_code);
 
         let options = metal::CompileOptions::new();
-        options.set_language_version(MTLLanguageVersion::V1_1);
+        options.set_language_version(MTLLanguageVersion::V1_2);
```
Comment (msiglreith) [inline src/backend/metal/src/device.rs]: Does this affect support for older devices? Do we need a fallback?

File: src/backend/metal/src/device.rs [hunk 4]
```diff
@@ -320,14 +333,22 @@ impl Device {
         } else {
             pipeline_desc.shaders.vertex.entry
         };
+
+        let vs_constants = if pipeline_desc.shaders.vertex.specialization.len() == 0 {
```
Comment (msiglreith) [inline src/backend/metal/src/device.rs]: nit: `is_empty()` instead of `len() == 0`

File: src/backend/metal/src/device.rs [hunk 5]
```diff
@@ -338,14 +359,22 @@ impl Device {
...
                 }
             };
+
             let fs_entry = if fs_remapped_entries.contains_key(fragment_entry.entry) {
                 fs_remapped_entries.get(fragment_entry.entry).unwrap()
             } else {
                 fragment_entry.entry
             };
+
+            let fs_constants = if fragment_entry.specialization.len() == 0 {
```
Comment (msiglreith) [inline src/backend/metal/src/device.rs]: (same as above)

File: src/backend/metal/src/device.rs (hunk 3 above)
Comment (grovesNL) [reply src/backend/metal/src/device.rs]: I'm not sure, I've had trouble locating a comprehensive compatibility table for the Metal language version.

Eventually we'll probably have to request the highest version we can and use supported capability flags, like the OpenGL backend had. We could only request `V1_2` when a specialization constant value is provided, but I don't know if it's worthwhile at the moment.

Comment (grovesNL) [reply src/backend/metal/src/device.rs]: https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf is the table I've been looking at so far, but it lists OS versions only. I don't know much about device compatibility with the various iOS/tvOS/macOS versions, or whether we have a minimum target for OS version.

**Comment by grovesNL** [comment]
Updated to address feedback, thanks!

File: src/backend/metal/src/device.rs (hunk 3 above)
Comment (msiglreith) [reply src/backend/metal/src/device.rs]: It seems to require a newer OS version at least:
https://developer.apple.com/documentation/metal/mtllanguageversion/1515361-version1_1
https://developer.apple.com/documentation/metal/mtllanguageversion/1639961-version1_2

**Comment by kvark** [comment]
@grovesNL https://crates.io/crates/metal-rs/0.6.0
Also note `cocoa` updated to 0.13, so you might want to follow here as well.

**Comment by grovesNL** [comment]
Updated `metal-rs` and `cocoa`, should be good to merge if we are okay with the version change for now.

**Comment by kvark** [comment]
Oh, totally!
Thanks :)
bors r+

**Pull request merged**